src/D2L.CodeStyle.TestAnalyzers/ServiceLocator/CustomTestServiceLocatorAnalyzer.cs

```diff
@@ -0,0 +1,152 @@
+			_excludeKnownProblems = excludeKnownProblemFixtures;
+		}
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction(
+				RegisterServiceLocatorAnalyzer
+			);
+		}
+
+		public void RegisterServiceLocatorAnalyzer(
+			CompilationStartAnalysisContext context
+		) {
+			INamedTypeSymbol factoryType = context.Compilation
+				.GetTypeByMetadataName( TestServiceLocatorFactoryType );
+
+			if( factoryType == null || factoryType.Kind == SymbolKind.ErrorType ) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(
+				ctx => PreventCustomLocatorUsage(
+					ctx,
+					factoryType
+				),
+				SyntaxKind.InvocationExpression
+			);
+		}
+
+		// Prevent static usage of TestServiceLocator.Create() methods.
+		private void PreventCustomLocatorUsage(
+			SyntaxNodeAnalysisContext context,
+			INamedTypeSymbol disallowedType
+		) {
+			ExpressionSyntax root = context.Node as InvocationExpressionSyntax;
+			if( root == null ) {
+				return;
+			}
+
+			SymbolInfo symbolInfo = context.SemanticModel.GetSymbolInfo( root );
+
+			IMethodSymbol method = symbolInfo.Symbol as IMethodSymbol;
+
+			if( method == null ) {
+				if( symbolInfo.CandidateSymbols == null ) {
+					return;
+				}
+
+				if( symbolInfo.CandidateSymbols.Length != 1 ) {
+					return;
+				}
+
+				// This happens on method groups, such as
+				// Func<IFoo> fooFunc = OldAndBrokenServiceLocator.Get<IFoo>;
+				method = symbolInfo.CandidateSymbols.First() as IMethodSymbol;
+
+				if( method == null ) {
+					return;
+				}
+			}
+
+			// If we're a Create method on a class that isn't
+			// TestServiceLocatorFactory, we're safe.
+			if( !IsTestServiceLocatorFactory(
+					actualType: method.ContainingType,
+					disallowedType: disallowedType
+				)
+			) {
+				return;
+			}
+
+			// If we're not a Create method, we're safe.
+			if( !IsCreateMethod( method ) ) {
+				return;
+			}
+
+			// Check whether any parent classes are on our whitelist. Checking
+			// all of the parents lets us handle partial classes more easily.
+			var parentClasses = context.Node.Ancestors().Where(
+				a => a.IsKind( SyntaxKind.ClassDeclaration )
+			);
+
+			var parentSymbols = parentClasses.Select(
+				c => context.SemanticModel.GetDeclaredSymbol( c )
+			);
+
+			if( parentSymbols.Any( IsClassWhitelisted ) ) {
+				return;
+			}
+
+			context.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.CustomServiceLocator,
+					context.Node.GetLocation()
+				)
+			);
+		}
+
+		private static bool IsTestServiceLocatorFactory(
+			INamedTypeSymbol actualType,
+			INamedTypeSymbol disallowedType
+		) {
+			if( actualType == null ) {
+				return false;
+			}
+
+			bool isLocatorFactory = actualType.Equals( disallowedType );
+			return isLocatorFactory;
+		}
+
+		private static bool IsCreateMethod(
+			IMethodSymbol method
+		) {
+			return method.Name.Equals( "Create" )
+				&& method.Parameters.Length > 0;
+		}

```

Comment:
So I worry about this guy a bit. Specifically, one use case of these custom locators is to mock out interfaces like `IEventNotifier`, `IConfigManager`, etc., that have undesirable side effects. Sometimes these are easy to fix (i.e. the offending interface is a direct dependency of the class being tested), but other times those dependencies run substantially deeper.

If we feel like this is setting too high a bar, one idea I had was to differentiate between usages of these methods that set a non-null (non-default) `dependencyLoader` parameter (warning in this case), and throwing an error in other cases. i.e. There's always a better solution to `Create( orgId, orgUnitId, userId )`, but that's not necessarily true if we're overloading a dependency.

Reply:
Yeah. So the current impl of custom object activators is just so horrible that we need to rid of all users. It's true that there are legitimateish usages like you describe. On the other hand... unit tests are better than big integration tests. Definitely some stuff to think about...

Reply:
Yeah, I think maybe it's not the worst thing in the world if we ruffle some feathers by banning them outright - adding elements to the whitelist isn't hard, and if we can make people think about whether there's a better way to do what they want to do, that's probably better?